subsys/usb/device_next/class/usbd_cdc_ncm.c

```diff
@@ -0,0 +1,1389 @@
+			connected ? "connected" : "disconnected", ret);
+	}
+
+	return ret;
+}
+
+static int cdc_ncm_send_speed_change(const struct device *dev)
+{
+	struct cdc_ncm_eth_data *data = dev->data;
+	struct usbd_class_data *c_data = data->c_data;
+	struct usbd_context *uds_ctx = usbd_class_get_ctx(c_data);
+	uint32_t usb_speed = (usbd_bus_speed(uds_ctx) == USBD_SPEED_FS) ?
+		NCM_USB_SPEED_FS : NCM_USB_SPEED_HS;
+	struct ncm_notify_connection_speed_change notify_speed_change = {
+		.header = {
+			.RequestType = {
+				.recipient = USB_REQTYPE_RECIPIENT_INTERFACE,
+				.type      = USB_REQTYPE_TYPE_CLASS,
+				.direction = USB_REQTYPE_DIR_TO_HOST
+			},
+			.bRequest = CONNECTION_SPEED_CHANGE,
+			.wLength  = sys_cpu_to_le16(8),
+			.wIndex = sys_cpu_to_le16(cdc_ncm_get_ctrl_if(data)),
+		},
+		.downlink = sys_cpu_to_le32(usb_speed),
+		.uplink   = sys_cpu_to_le32(usb_speed),
+	};
+	int ret;
+
+	ret = cdc_ncm_send_notification(dev,
+					&notify_speed_change,
+					sizeof(notify_speed_change));
+	if (ret < 0) {
+		LOG_DBG("Cannot send %s (%d)", "speed change", ret);
+		return ret;
+	}
+
+	return ret;
+}
+
+
+static int ncm_send_notification(const struct device *dev)
+{
+	struct cdc_ncm_eth_data *data = dev->data;
+	int ret;
+
+	/* Speed change must be sent first, chapter 7.1 */
+	if (data->if_state == IF_STATE_INIT) {
+		ret = cdc_ncm_send_speed_change(dev);
+		if (ret < 0) {
+			LOG_DBG("Cannot send %s (%d)", "speed change", ret);
+			return ret;
+		}
+
+		LOG_DBG("Speed change sent");
+		data->if_state = IF_STATE_SPEED_CHANGE_SENT;
+		return -EAGAIN;
+	}
+
+	if (data->if_state == IF_STATE_SPEED_CHANGE_SENT) {
+		ret = cdc_ncm_send_connected(dev, true);
+		if (ret < 0) {
+			LOG_DBG("Cannot send %s (%d)", "connected status", ret);
+			return ret;
+		}
+
+		LOG_DBG("Connected status sent");
+		data->if_state = IF_STATE_CONNECTION_STATUS_SENT;
+		return -EAGAIN;
+	}
+
+	if (data->if_state == IF_STATE_CONNECTION_STATUS_SENT) {
+		data->if_state = IF_STATE_DONE;
+	}
+
+	return 0;
+}
+
+static void send_notification_work(struct k_work *work)
+{
+	struct k_work_delayable *notif_work = k_work_delayable_from_work(work);
+	struct cdc_ncm_eth_data *data;
+	struct device *dev;
+	int ret;
+
+	data = CONTAINER_OF(notif_work, struct cdc_ncm_eth_data, notif_work);
+	dev = usbd_class_get_private(data->c_data);
+
+	if (atomic_test_bit(&data->state, CDC_NCM_IFACE_UP)) {
+		ret = ncm_send_notification(dev);
+	} else {
+		ret = cdc_ncm_send_connected(dev, false);
+	}
+
+	if (ret) {
+		(void)k_work_reschedule(&data->notif_work, K_MSEC(100));
+	}
+}
+
+static void usbd_cdc_ncm_update(struct usbd_class_data *const c_data,
+				const uint8_t iface, const uint8_t alternate)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *data = dev->data;
+	struct usbd_cdc_ncm_desc *desc = data->desc;
+	uint8_t data_iface = desc->if1_1.bInterfaceNumber;
+	int ret;
+
+	LOG_WRN("New configuration, interface %u alternate %u",
+		iface, alternate);
```

Comment:
I think it should be LOG_INF, otherwise we will get bug reports for this.
```
	LOG_INF("New configuration, interface %u alternate %u",
		iface, alternate);
```